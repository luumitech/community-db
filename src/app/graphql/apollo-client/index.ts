import {
  ApolloClient,
  ApolloLink,
  FetchResult,
  HttpLink,
  InMemoryCache,
  Observable,
  Operation,
  from,
  split,
} from '@apollo/client';
import { onError } from '@apollo/client/link/error';
import { createPersistedQueryLink } from '@apollo/client/link/persisted-queries';
import { getMainDefinition } from '@apollo/client/utilities';
import createUploadLink from 'apollo-upload-client/createUploadLink.mjs';
import { print } from 'graphql';
import { Client, ClientOptions, createClient } from 'graphql-sse';
import { typePolicies } from './type-policies';

const GRAPHQL_URL = '/api/graphql';

const cache = new InMemoryCache({
  typePolicies,
});

/** GraphQL over SSE, for handling subscription */
class SSELink extends ApolloLink {
  private client: Client;

  constructor(options: ClientOptions) {
    super();
    this.client = createClient(options);
  }

  public request(operation: Operation): Observable<FetchResult> {
    return new Observable((sink) => {
      return this.client.subscribe<FetchResult>(
        {
          ...operation,
          // query: print(operation.query)
          // When using persisted query, information is embedded in
          // extension object
          query: '',
        },
        {
          next: sink.next.bind(sink),
          complete: sink.complete.bind(sink),
          error: sink.error.bind(sink),
        }
      );
    });
  }
}

/** Unified error handling for apollo */
const errorLink = onError((response) => {
  const { graphQLErrors, networkError, operation } = response;
  if (graphQLErrors) {
    graphQLErrors.map(({ message, locations, path, extensions }) =>
      console.warn(
        `[GraphQL error]: Message: ${message}, Path: ${path}, Location: %o, extensions: %o`,
        locations,
        extensions
      )
    );
  }
  if (networkError) {
    console.warn(networkError);
  }
  console.warn('operation: %o', operation);
});

const persistedQuerylink = createPersistedQueryLink({
  // @ts-expect-error __meta__ inserted by graphql yoga
  generateHash: (document) => document.__meta__.hash,
});

// Non-subscription enabled link
// const httpLink = new HttpLink({
const httpLink = createUploadLink({ uri: GRAPHQL_URL });

// Subscription enabled link
const sseLink = new SSELink({ url: GRAPHQL_URL });

const splitLink = split(
  ({ query }) => {
    const definition = getMainDefinition(query);
    return (
      definition.kind === 'OperationDefinition' &&
      definition.operation === 'subscription'
    );
  },
  from([sseLink]),
  from([errorLink, httpLink])
);

const apolloClient = new ApolloClient({
  link: from([persistedQuerylink, splitLink]),
  cache,
});

export default apolloClient;
