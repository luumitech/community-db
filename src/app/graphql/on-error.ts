import { ApolloError } from '@apollo/client';
import { toast } from '~/view/base/toastify';

interface OnErrorOpt {
  /**
   * Handle error on your own, if the error is not handled, return it, and it
   * will be handled by the default error handler
   */
  customOnError?: (err: ApolloError) => ApolloError | undefined | void;
}

export function onError(error: ApolloError, opt?: OnErrorOpt) {
  if (opt?.customOnError == null) {
    toast.error(error.message);
    return;
  }

  // Custom error handler
  const unhandledError = opt.customOnError(error);
  if (unhandledError) {
    // Default error goes to toast
    toast.error(unhandledError.message);
  }
}
