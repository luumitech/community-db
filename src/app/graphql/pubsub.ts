/**
 * Graphql-Yoga pubsub
 */
import { createPubSub } from '@graphql-yoga/subscription';
import type { Community, Property } from '@prisma/client';

// See: https://the-guild.dev/graphql/yoga-server/tutorial/advanced/02-subscriptions#subscribing-to-new-link-elements
export enum MutationType {
  CREATED = 'CREATED',
  UPDATED = 'UPDATED',
  DELETED = 'DELETED',
}

export interface PubSubEvent {
  /**
   * email (AuthContext) of broadcaster
   */
  broadcasterId: string;
  mutationType: MutationType;
}

export interface PubSubCommunityEvent extends PubSubEvent {
  community: Community;
}

export interface PubSubPropertyEvent extends PubSubEvent {
  property: Property;
}

interface PubSubEventsImpl {
  // i.e. /community/${communityId}/
  [key: `community/${string}/`]: [PubSubCommunityEvent];
  // i.e. /community/${communityId}/property
  [key: `community/${string}/property`]: [PubSubPropertyEvent];
}
type ValueOf<T> = T[keyof T];

type PubSubEvents = PubSubEventsImpl &
  Record<string, ValueOf<PubSubEventsImpl>>;

export const pubSub = createPubSub<PubSubEvents>();
