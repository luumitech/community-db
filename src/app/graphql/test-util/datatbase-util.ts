import { Prisma } from '@prisma/client';
import * as XLSX from 'xlsx';
import { importLcraDB } from '~/lib/lcra-community/import';
import prisma from '~/lib/prisma';

/** Database related utilities for testing purpose */
export class DatabaseUtil {
  /**
   * Seed database with excel file in `fixturePath`
   *
   * @param fixturePath Excel workbook
   * @returns Database entries returned
   */
  async seed(fixturePath: string) {
    const workbook = XLSX.readFile(fixturePath);
    const importResult = importLcraDB(workbook);
    const { propertyList, ...others } = importResult;
    const communitySeed: Prisma.CommunityCreateInput[] = [
      {
        name: 'Test Community',
        ...others,
        propertyList: {
          create: propertyList,
        },
      },
    ];

    const accessSeed: Prisma.AccessCreateWithoutUserInput[] = communitySeed.map(
      (community) => ({
        role: 'ADMIN',
        community: {
          create: community,
        },
      })
    );

    await prisma.user.create({
      data: {
        email: 'jest@email.com',
        accessList: {
          create: accessSeed,
        },
      },
    });

    return importResult;
  }
}
