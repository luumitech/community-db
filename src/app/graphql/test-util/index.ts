import { GraphQLUtil } from './graphql-util';

/**
 * Helper test utility class for initializaton of key
 * test tools when running Jest unit tests
 */
export class TestUtil {
  public graphql: GraphQLUtil;

  constructor() {
    this.graphql = new GraphQLUtil();
  }

  /**
   * Initialization routine for setting services required to run jest tests
   * - clears and initialize mongodb
   * - setup blob service
   *
   * This is expected to be constructed in `beforeAll` of each test suite
   * to initialize the mongoose connection
   */
  async initialize() {}

  /**
   * Termination routine to clean up services used during jest tests
   *
   * This is expected to be constructed in `afterAll` of each test suite
   * to terminate the mongoose connection
   */
  async terminate() {}
}
