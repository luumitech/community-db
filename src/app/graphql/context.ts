import { GraphQLError } from 'graphql';
import { type YogaInitialContext } from 'graphql-yoga';
import { getServerSession } from 'next-auth/next';
import { authOptions } from '~/api/auth/[...nextauth]/auth-options';
import prisma from '../lib/prisma';
import { pubSub } from './pubsub';

export async function createContext(ctx: YogaInitialContext) {
  const session = await getServerSession(authOptions);

  // All graphQL operations require user to be authenticated
  if (!session) {
    throw new GraphQLError('You are not authenticated');
  }

  const { user } = session;
  if (!user) {
    throw new GraphQLError('Auth: Missing user context');
  }
  const { uid } = user;
  if (!uid) {
    throw new GraphQLError('Auth: Missing uid in user context');
  }
  return {
    user,
    pubSub,
  };
}

export type Context = Awaited<ReturnType<typeof createContext>>;
