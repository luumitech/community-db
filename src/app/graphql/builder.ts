import SchemaBuilder from '@pothos/core';
import PrismaPlugin from '@pothos/plugin-prisma';
import RelayPlugin from '@pothos/plugin-relay';
import { DateResolver, DateTimeResolver } from 'graphql-scalars';
import prisma from '../lib/prisma';
import { createContext } from './context';
import type PrismaTypes from './generated/pothos-types.ts';

export const builder = new SchemaBuilder<{
  PrismaTypes: PrismaTypes;
  Context: ReturnType<typeof createContext>;
  Scalars: {
    Date: { Input: Date; Output: Date };
    DateTime: { Input: Date; Output: Date };
  };
  DefaultEdgesNullability: false;
}>({
  plugins: [PrismaPlugin, RelayPlugin],
  relayOptions: {
    cursorType: 'String',
    edgesFieldOptions: {
      nullable: false,
    },
  },
  prisma: {
    client: prisma,
  },
});

builder.addScalarType('Date', DateResolver, {});
builder.addScalarType('DateTime', DateTimeResolver, {});
builder.queryType({});
builder.mutationType({});
