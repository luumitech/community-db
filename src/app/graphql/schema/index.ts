import { builder } from '../builder';

import './access';
import './community';
import './geocode';
import './offset-pagination';
import './payment';
import './property';
import './subscription';
import './user';

export const schema = builder.toSchema();
