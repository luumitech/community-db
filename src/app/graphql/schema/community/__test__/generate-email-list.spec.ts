import path from 'path';
import { graphql } from '~/graphql/generated';
import { TestUtil } from '~/graphql/test-util';

describe('Generate Email List', () => {
  const testUtil = new TestUtil();

  beforeAll(async () => {
    await testUtil.initialize();
    await testUtil.database.seed(
      path.join(process.cwd(), '__fixtures__', 'simple.xlsx')
    );
  });

  afterAll(async () => {
    await testUtil.terminate();
  });

  test('Verify properties that have not been renewed', async () => {
    const document = graphql(/* GraphQL */ `
      query GenerateEmailList_NoRenewalPropertyList {
        userCurrent {
          accessList {
            community {
              communityStat {
                noRenewalPropertyList(year: 2024) {
                  address
                  occupantList {
                    email
                  }
                }
              }
            }
          }
        }
      }
    `);

    const result = await testUtil.graphql.executeSingle({ document });
    expect(result.data).toMatchSnapshot();
  });

  test('Verify properties that are members', async () => {
    const document = graphql(/* GraphQL */ `
      query GenerateEmailList_EmailMemberPropertyList {
        userCurrent {
          accessList {
            community {
              communityStat {
                memberPropertyList(year: 2022) {
                  address
                  occupantList {
                    email
                  }
                }
              }
            }
          }
        }
      }
    `);

    const result = await testUtil.graphql.executeSingle({ document });
    expect(result.data).toMatchSnapshot();
  });

  test('Verify properties that are NOT members', async () => {
    const document = graphql(/* GraphQL */ `
      query GenerateEmailList_NonMemberPropertyList {
        userCurrent {
          accessList {
            community {
              communityStat {
                nonMemberPropertyList(year: 2022) {
                  address
                  occupantList {
                    email
                  }
                }
              }
            }
          }
        }
      }
    `);

    const result = await testUtil.graphql.executeSingle({ document });
    expect(result.data).toMatchSnapshot();
  });
});
