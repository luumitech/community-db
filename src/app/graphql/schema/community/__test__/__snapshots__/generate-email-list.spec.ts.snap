// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate Email List Verify properties that are NOT members 1`] = `
{
  "__typename": "Query",
  "userCurrent": {
    "__typename": "User",
    "accessList": [
      {
        "__typename": "Access",
        "community": {
          "__typename": "Community",
          "communityStat": {
            "__typename": "CommunityStat",
            "nonMemberPropertyList": [
              {
                "__typename": "Property",
                "address": "123 Adventure Drive",
                "occupantList": [
                  {
                    "__typename": "Occupant",
                    "email": "ksmith@email.com",
                  },
                ],
              },
            ],
          },
        },
      },
    ],
  },
}
`;

exports[`Generate Email List Verify properties that are members 1`] = `
{
  "__typename": "Query",
  "userCurrent": {
    "__typename": "User",
    "accessList": [
      {
        "__typename": "Access",
        "community": {
          "__typename": "Community",
          "communityStat": {
            "__typename": "CommunityStat",
            "memberPropertyList": [
              {
                "__typename": "Property",
                "address": "99 Fortune Drive",
                "occupantList": [
                  {
                    "__typename": "Occupant",
                    "email": "Email1",
                  },
                  {
                    "__typename": "Occupant",
                    "email": "Email2",
                  },
                  {
                    "__typename": "Occupant",
                    "email": "Email3",
                  },
                ],
              },
            ],
          },
        },
      },
    ],
  },
}
`;

exports[`Generate Email List Verify properties that have not been renewed 1`] = `
{
  "__typename": "Query",
  "userCurrent": {
    "__typename": "User",
    "accessList": [
      {
        "__typename": "Access",
        "community": {
          "__typename": "Community",
          "communityStat": {
            "__typename": "CommunityStat",
            "noRenewalPropertyList": [
              {
                "__typename": "Property",
                "address": "123 Adventure Drive",
                "occupantList": [
                  {
                    "__typename": "Occupant",
                    "email": "ksmith@email.com",
                  },
                ],
              },
              {
                "__typename": "Property",
                "address": "99 Fortune Drive",
                "occupantList": [
                  {
                    "__typename": "Occupant",
                    "email": "Email1",
                  },
                  {
                    "__typename": "Occupant",
                    "email": "Email2",
                  },
                  {
                    "__typename": "Occupant",
                    "email": "Email3",
                  },
                ],
              },
            ],
          },
        },
      },
    ],
  },
}
`;
